### PR TITLE
Multiple EventPipe related bugs/issues

### PR DESCRIPTION
**Pull request by jorive**
- Addressing https://github.com/dotnet/coreclr/pull/24896#discussion_r291790885
- EventPipe::Shutdown can be called concurrently from multiple threads, then EventPipe::s_pSessions could change from not-null to null, which could trigger an AV.
- Addressing the following assertion:
```log
CoreRun.exe ".\bin\tests\Windows_NT.x64.Debug\tracing\runtimeeventsource\runtimeeventsource\runtimeeventsource.exe"
-------------------------------------------------------------------------------


Assert failure(PID 25672 [0x00006448], Thread: 29384 [0x72c8]): CONTRACT VIOLATION by EventPipeBufferManager::MoveNextEventAnyThread at ".\src\vm\eventpipebuffermanager.cpp" @ 755

THROWS called in a NOTHROW region.

                        CONTRACT in EventPipeBufferManager::MoveNextEventAnyThread at ".\src\vm\eventpipebuffermanager.cpp" @ 755
VIOLATED-->  CONTRACT in EventPipeBufferManager::GetNextEvent at ".\src\vm\eventpipebuffermanager.cpp" @ 705
                        CONTRACT in EventPipeSession::GetNextEvent at ".\src\vm\eventpipesession.cpp" @ 385
                        CONTRACT in EventPipe::GetNextEvent at ".\src\vm\eventpipe.cpp" @ 889
                        CONTRACT in EventPipeInternal::GetNextEvent at ".\src\vm\eventpipeinternal.cpp" @ 264
                        OVERRIDE_TYPE_LOAD_LEVEL_LIMIT in MethodDescCallSite::CallTargetWorker at ".\src\vm\callhelpers.cpp" @ 354
                        CONTRACT in MethodDescCallSite::CallTargetWorker at ".\src\vm\callhelpers.cpp" @ 344
                        CONTRACT in ThreadNative::KickOffThread_Worker at ".\src\vm\comsynchronizable.cpp" @ 193
                        CONTRACT in ManagedThreadBase_DispatchInner at ".\src\vm\threads.cpp" @ 7407
                        CONTRACT in ManagedThreadBase_FullTransition at ".\src\vm\threads.cpp" @ 7663
                        CONTRACT in ThreadNative::KickOffThread at ".\src\vm\comsynchronizable.cpp" @ 289



CORECLR! CONTRACT_ASSERT + 0x426 (0x00007ffc`d2473e16)
CORECLR! BaseContract::DoChecks + 0x233 (0x00007ffc`d24766c7)
CORECLR! EEContract::DoChecks + 0xD3 (0x00007ffc`d253de57)
CORECLR! EventPipeBufferManager::MoveNextEventAnyThread + 0x2CA (0x00007ffc`d2cd8316)
CORECLR! EventPipeBufferManager::GetNextEvent + 0xC5 (0x00007ffc`d2cd5aa5)
CORECLR! EventPipeSession::GetNextEvent + 0xB8 (0x00007ffc`d2b9d044)
CORECLR! EventPipe::GetNextEvent + 0xC4 (0x00007ffc`d282b50c)
CORECLR! EventPipeInternal::GetNextEvent + 0x162 (0x00007ffc`d2c3d952)
<no module>! <no symbol> + 0x0 (0x00007ffc`73305484)
<no module>! <no symbol> + 0x0 (0x00000216`57e4bb00)
    File: .\src\vm\eventpipebuffermanager.cpp Line: 755
    Image: .\bin\tests\Windows_NT.x64.Checked\Tests\Core_Root\CoreRun.exe
```
- Removed unused `EventPipe::s_fileSwitchTimerHandle` static member.

### CODE REVIEW

File: src/vm/ceemain.cpp [hunk 1]
```diff
@@ -1290,9 +1290,11 @@ void STDMETHODCALLTYPE EEShutDownHelper(BOOL fIsDllUnloading)
     }
 
 #ifdef FEATURE_PERFTRACING
-    // Shutdown the event pipe.
-    EventPipe::Shutdown();
-    DiagnosticServer::Shutdown();
+    if (!fIsDllUnloading)
```
Comment (jorive) [inline src/vm/ceemain.cpp]: @brianrob would you know why `ETW::EnumerationLog::ProcessShutdown();` is called when `fIsDllUnloading` (the second time Shutdown is called). Isn't it dangerous to do cleanup there, am I missing something here? Could you please give us some background here on what's appropriate?

File: src/vm/eventpipe.cpp [hunk 2]
```diff
@@ -185,6 +184,10 @@ bool EventPipeProviderCallbackDataQueue::TryDequeue(EventPipeProviderCallbackDat
 void EventPipe::Initialize()
 {
     STANDARD_VM_CONTRACT;
+    PRECONDITION(!s_tracingInitialized);
+
+    if (s_tracingInitialized)
```
Comment (sywhang) [inline src/vm/eventpipe.cpp]: Is it safe to check this without taking the lock?

File: src/vm/eventpipe.cpp [hunk 3]
```diff
@@ -238,28 +240,37 @@ void EventPipe::Shutdown()
...
+            for (uint32_t i = 0; i < MaxNumberOfSessions; ++i)
+            {
+                EventPipeSession *pSession = s_pSessions[i].Load();
+                if (pSession)
+                    Disable(static_cast<EventPipeSessionID>(1ULL << i));
+            }
+
+            // Remove EventPipeEventSource first since it tries to use the data structures that we remove below.
+            // We need to do this after disabling sessions since those try to write to EventPipeEventSource.
+            delete s_pEventSource;
```
Comment (sywhang) [inline src/vm/eventpipe.cpp]: Is it safe to do this here without taking the EventPipeCrst lock?

File: src/vm/eventpipe.cpp (hunk 2 above)
Comment (jorive) [reply src/vm/eventpipe.cpp]: Need to review all places where we use the flag.

Comment (josalem) [reply src/vm/eventpipe.cpp]: Does your comment mean this function can be entered more than once?  If this function can be entered multiple times, won't `s_tracingInitialized` be true on later entries, thus violating the precondition?

File: src/vm/eventpipe.cpp [hunk 4]
```diff
@@ -293,6 +301,9 @@ EventPipeSessionID EventPipe::Enable(
 
     EventPipeSessionID sessionId = 0;
     RunWithCallbackPostponed([&](EventPipeProviderCallbackDataQueue *pEventPipeProviderCallbackDataQueue) {
+        if (!s_tracingInitialized)
```
Comment (josalem) [inline src/vm/eventpipe.cpp]: Just curious: What does moving this check inside the callback change?

Comment (jorive) [reply src/vm/eventpipe.cpp]: Inside the callback, the check is protected by the Crst.

File: src/vm/eventpipe.cpp (hunk 2 above)
Comment (jorive) [reply src/vm/eventpipe.cpp]: It should not be entered more than once, but I wanted to test it, as Shutdown was called twice.

**Comment by jorive** [comment]
For more information, these were the two call stacks when `EventPipe::Shutdown` was called:
```log
First call (fIsDllUnloading=0)
>	coreclr.dll!EventPipe::Shutdown() Line 219	C++
 	coreclr.dll!EEShutDownHelper(int fIsDllUnloading) Line 1295	C++
 	coreclr.dll!EEShutDown(int fIsDllUnloading) Line 1776	C++
 	coreclr.dll!CorHost2::UnloadAppDomain2(unsigned long dwDomainId, int fWaitUntilDone, int * pLatchedExitCode) Line 1035	C++
 	CoreRun.exe!TryRun(const int argc, const wchar_t * * argv, Logger & log, const bool verbose, const bool waitForDebugger, unsigned long & exitCode) Line 713	C++
 	CoreRun.exe!wmain(const int argc, const wchar_t * * argv) Line 813	C++

Second call (fIsDllUnloading!=0)
>	coreclr.dll!EventPipe::Shutdown() Line 219	C++
 	coreclr.dll!EEShutDownHelper(int fIsDllUnloading) Line 1295	C++
 	coreclr.dll!EEShutDown(int fIsDllUnloading) Line 1776	C++
 	coreclr.dll!`EEDllMain'::`5'::__Body::Run(EEDllMain::__l2::Param * pParam) Line 1946	C++
 	coreclr.dll!EEDllMain(HINSTANCE__ * hInst, unsigned long dwReason, void * lpReserved) Line 1995	C++
 	coreclr.dll!DllMain(void * hInstance, unsigned long dwReason, void * lpReserved) Line 148	C++
 	coreclr.dll!CoreDllMain(void * hInstance, unsigned long dwReason, void * lpReserved) Line 107	C++
 	[External Code]	
 	CoreRun.exe!exit_or_terminate_process(const unsigned int return_code) Line 144	C++
 	CoreRun.exe!common_exit(const int return_code, const _crt_exit_cleanup_mode cleanup_mode, const _crt_exit_return_mode return_mode) Line 282	C++
 	CoreRun.exe!exit(int return_code) Line 294	C++
 	[External Code]	
```

File: src/vm/eventpipe.cpp (hunk 2 above)
Comment (sywhang) [reply src/vm/eventpipe.cpp]: Nvm. `s_tracingInitialized` is volatile, so I'm guessing this is fine.

File: src/vm/eventpipe.cpp [hunk 5]
```diff
@@ -640,10 +650,6 @@ void EventPipe::WriteEventInternal(EventPipeEvent &event, EventPipeEventPayload
     }
     CONTRACTL_END;
 
-    // We can't proceed if tracing is not initialized.
-    if (!s_tracingInitialized)
```
Comment (noahfalk) [inline src/vm/eventpipe.cpp]: I would leave these checks in place until we can replace it with something that is more rigorously correct. With this check  WriteEvent will fail in a relatively narrow race with Shutdown that deletes event before it is used below. Without this check this method will AV anytime Write is called after Shutdown.

Comment (jorive) [reply src/vm/eventpipe.cpp]: We can take care of the pre-existing issues later then.

**Pull request merged**